src/mongo-utils.ts

```diff
@@ -6,28 +6,58 @@ import * as mongodb from 'mongodb';
 
 import { LodashReplacerUtils } from './lodash-replacer.utils';
 import { ClassType } from './models/class-type.models';
+import { MongoUtilsOptions } from './models/mongo-utils-options.models';
 
 export class MongoUtils {
 	private static readonly MONGO_ID_REGEXP: RegExp = /^[0-9a-f]{24}$/;
 
 	public static async connect(
 		url: string,
 		options: mongodb.MongoClientOptions = { useNewUrlParser: true },
+		connectOptions: MongoUtilsOptions = { killProcessOnReconnectFailed: true },
 	): Promise<mongodb.Db> {
 		const log = global.log.module('mongo');
 		log.info(`Connecting to ${MongoUtils.hidePasswordFromURI(url)}...`);
-		global.dbClient = await mongodb.MongoClient.connect(url, options);
-		const db = (global.dbClient as mongodb.MongoClient).db();
+
+		const mongoClient = new mongodb.MongoClient(url, options);
+		const safeUrl = MongoUtils.hidePasswordFromURI(url);
+
+		mongoClient.on('serverOpening', () => {
+			log.info(`Connection to the mongodb server is being established...`);
+		});
+		mongoClient.on('open', () => {
+			log.info(`Client connected to ${safeUrl}.`);
+		});
+		mongoClient.on('close', () => {
+			log.info(`Client disconnected from ${safeUrl}.`);
+		});
+		mongoClient.on('reconnect', () => {
+			log.info(`Client reconnected to ${safeUrl}.`);
+		});
+		mongoClient.on('timeout', () => {
+			log.warn(`Client connection or operation timed out`);
+		});
+
+		const dbClient = await mongoClient.connect();
+		global.dbClient = dbClient;
+		const db = dbClient.db();
 		global.db = db;
-		log.info(`Connected`);
+
+		(db as any).s.topology.on('reconnectFailed', () => {
+			log.warn(`Client reconnection failed.`);
+			if (connectOptions.killProcessOnReconnectFailed) {
+				process.kill(process.pid, 'SIGTERM');
```

Comment:
new idea, have you tried to mock process.kill with mockito ?

Reply:
yes, i tried to mock process.kill using the following code
```typescript
const processMock: NodeJS.Process = Mockito.mock(process);
Mockito.when(processMock.kill(Mockito.anything(), Mockito.anything())).thenReturn();
```
However, I end up with an error saying `processMock.kill is not a function`.
Looking it up online, it seems that the NodeJS.Process cannot be easily mocked...

An alternative way would be to create a new file 'reconnectFailed.js' that runs the code that will call process.kill, and run it in a separate child process...

```typescript
const childProcess = spawn('node', ['reconnectFailed.js']);
await waitFor(2000); // Adjust the time as needed
t.true(childProcess.killed);
```
What's your take on it?

Reply:
seems overkill to me, leave it that way, we tried :)